t/ui/21-admin-needles.t

```diff
@@ -111,6 +112,58 @@ like(
     "redirected to right module too"
 );
 
+subtest 'dereference symlink when displaying needles info' => sub {
+    my $real_needle_dir    = getcwd . '/t/data/openqa/share/tests/opensuse';
+    my $symlink_needle_dir = getcwd . '/t/data/openqa/share/tests/test_symlink_dir';
+    symlink($real_needle_dir, $symlink_needle_dir) or die "cannot make soft link $!";
+    my $needle_dir            = $schema->resultset('NeedleDirs');
+    my $symlink_needle_dir_id = $needle_dir->create(
+        {
+            path => $symlink_needle_dir . '/needles',
+            name => 'symlink_needle_dir'
+        })->id;
+    my $real_needle_dir_id = $needle_dir->create(
+        {
+            path => $real_needle_dir . '/needles',
+            name => 'real_needle_dir'
+        })->id;
+    my $needles        = $schema->resultset('Needles');
+    my $symlink_needle = $needles->create(
+        {
+            dir_id                 => $symlink_needle_dir_id,
+            filename               => 'bootloader.json',
+            last_seen_module_id    => undef,
+            last_seen_time         => undef,
+            last_matched_module_id => undef,
+            file_present           => 1,
+        });
+    my $real_needle = $needles->create(
+        {
+            dir_id                 => $real_needle_dir_id,
+            filename               => 'bootloader.json',
+            last_seen_module_id    => 10,
+            last_seen_time         => time2str('%Y-%m-%d %H:%M:%S', time - 100000),
+            last_matched_module_id => 9,
+            last_matched_time      => time2str('%Y-%m-%d %H:%M:%S', time - 50000),
+            file_present           => 1,
+            t_created              => time2str('%Y-%m-%d %H:%M:%S', time - 200000),
+            t_updated              => time2str('%Y-%m-%d %H:%M:%S', time - 200000),
+        });
+
+    goto_admin_needle_table();
+    foreach my $tr (@{$driver->find_elements('#needles tbody tr')}) {
```

Comment:
I wouldn't repeat `$tr->children('td')` all the time (especially inside a loop).

I assume it would be an error if the row is completely missing. Unfortunately it seems like there's no check to verify whether the row is present at all.